test/demo.jl

```diff
@@ -42,15 +48,31 @@ end
 @testset "Demo ComponentsArrays" begin
 
     # target distribution parametrized by ComponentsArray
-    p1 = ComponentVector(μ = 2.0, σ=1)
-    ℓπ(p::ComponentArray) = -(1 .- p.μ)^2/(p.σ)^2
+    p1 = ComponentVector(μ=2.0, σ=1)
+    struct DemoProblemComponentArrays end
+    function LogDensityProblems.logdensity(::DemoProblemComponentArrays, p::ComponentArray)
+        return -((1 - p.μ) / p.σ)^2
+    end
+    LogDensityProblems.dimension(::DemoProblemComponentArrays) = 2
+    ℓπ = DemoProblemComponentArrays()
 
     # Define a Hamiltonian system
     D = length(p1)          # number of parameters
     metric = DiagEuclideanMetric(D)
 
     # choose AD framework or provide a function manually
-    hamiltonian = Hamiltonian(metric, ℓπ, ForwardDiff)
+    ℓπ_with_grad = LogDensityProblemsAD.ADgradient(
+        Val(:ForwardDiff),
+        ℓπ;
+        # NOTE: Have to manually construct the `gradientconfig` to ensure we're not using the default
+        # buffer in the config (which won't be of type `ComponentArray`).
+        gradientconfig = ForwardDiff.GradientConfig(
+            Base.Fix1(LogDensityProblems.logdensity, ℓπ),
+            similar(p1),
+            LogDensityProblemsAD._default_chunk(ℓπ),
+        )
```

Comment:
As you said, that's not convenient for users. Could we just add a method
```julia
Hamiltonian(metric, logdensity, ::Val{:ForwardDiff}; kwargs...)
```
and perform the gradient construction with the correct config there automatically? With the `kwargs` we could allow some customization by users.

Generally, I guess it would be useful to forward `kwargs` to `ADgradient` also for other backends.

Reply:
Generally, maybe that should also be fixed upstream: Similar to the ReverseDiff options (https://github.com/tpapp/LogDensityProblemsAD.jl/blob/79e245e1bb8e904087af9de284e2796e51eb55c4/src/AD_ReverseDiff.jl#L32), one could add a `x` keyword argument (default: `nothing`) that would allow to specify the type of input vector.

Reply:
You can't specify the underlying type used in the `GradientConfig` using a `kwarg` for `ADgradient` though :confused: That is, we the "best" simplification we can do (without upstream changes) is

```julia
Hamiltonian(metric, logdensity, ::Val{:ForwardDiff}; gradientconfig)
```

and the user still has to construct `gradientconfig` as above.

We can of course also make some helper method for constructing `GradientConfig` for a particular type, or, maybe better, add an additional argument to `LogDensityProblemsAD.ADgradient` which is passed to `LogDensityProblemsAD._default_gradientconfig`.

Reply:
> maybe better, add an additional argument to LogDensityProblemsAD.ADgradient which is passed to LogDensityProblemsAD._default_gradientconfig.

That's what I would suggest :slightly_smiling_face: 

> one could add a x keyword argument (default: nothing) that would allow to specify the type of input vector.

However, even if that's not available yet - we can use in `Hamiltonian` whatever keyword arguments we want, can't we? So we could add such a `x` keyword argument at least there right away and then define a gradient config based on it.

Reply:
> That's what I would suggest

https://github.com/tpapp/LogDensityProblemsAD.jl/pull/3

> However, even if that's not available yet - we can use in Hamiltonian whatever keyword arguments we want, can't we? So we could add such a x keyword argument at least there right away and then define a gradient config based on it.

Yeah, but if @tpapp is okay with the PR me might as well just use that to avoid (potentially) more breaking changes, e.g. if we choose a diff kwarg than LogDensityProblemsAD does.

Reply:
I am OK with the PR, conditional on changes requested by @devmotion.